nixos/syncoid: enable N:N dataset mappings

ju1m opened the pull request:
###### Description of changes
Same as https://github.com/NixOS/nixpkgs/pull/147559 but rebased onto `master` instead of `release-23.05`.
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

ju1m committed:
nixos/sanoid: add utilities useful to syncoid

ju1m committed:
nixos/syncoid: use DynamicUser=

The following changes were previously in separate commits
but have been squashed to be able to rebase them
onto a treewide change applying nixfmt-rfc-style
on nixos/modules/services/backup/syncoid.nix
which introduced multiple untrackable conflicts.

Changes:

- nixos/syncoid: add destroy to localTargetAllow's default

syncoid[3282027]: Resuming interrupted zfs send/receive from rpool/var/mail to losurdo/backup/mermet/var/mail (~ UNKNOWN remaining):
syncoid[3282885]: cannot resume send: 'rpool/var/mail@autosnap_2021-10-17_15:00:19_hourly' used in the initial send no longer exists
syncoid[3282897]: cannot receive: failed to read from stream
syncoid[3282027]: WARN: resetting partially receive state because the snapshot source no longer exists
syncoid[3283326]: cannot destroy 'losurdo/backup/mermet/var/mail/%recv': permission denied

- nixos/syncoid: allow @timer syscalls

- nixos/syncoid: set TZ envvar

- nixos/syncoid: test sending to multiple targets

- nixos/syncoid: don't delegate permissions to source's parent
See https://github.com/NixOS/nixpkgs/pull/165204
and https://github.com/NixOS/nixpkgs/pull/180111

- nixos/syncoid: add support for LoadCredentialEncrypted=

- nixos/syncoid: zfs-unallow unused dynamic users

- nixos/syncoid: use NFTSet=

lopsided98 commented on nixos/modules/services/backup/sanoid.nix:
```diff
@@ -199,6 +199,13 @@ in
+
+    # Put those packages in the global environment
+    # so that syncoid can find them when targeting this host through ssh.
+    environment.systemPackages = with pkgs; [
```
I guess this is mostly harmless if they aren't used, but I'm not sure it makes sense to assume anyone running sanoid will also be using the machine as a syncoid target. Personally, I only add these packages for a single user on my target machines, just because it seems tidier.

lopsided98 commented on nixos/modules/services/backup/sanoid.nix:
```diff
@@ -199,6 +199,13 @@ in
+
+    # Put those packages in the global environment
+    # so that syncoid can find them when targeting this host through ssh.
+    environment.systemPackages = with pkgs; [
```
I guess this is mostly harmless if they aren't used, but I'm not sure it makes sense to assume anyone running sanoid will also be using the machine as a syncoid target. Personally, I only add these packages for a single user on my target machines, just because it seems tidier.

ju1m replied:
Yes, you're both right, and IIRC `syncoid` outputs a warning message when those (quite small) utilities are missing (`lzop`, `mbuffer`), yet I think it's quite easy to forget about installing them. Hosts running `sanoid` are likely to be the targets of tools using them (be it `syncoid` or another one), that's why I think it's a good thing to made them available.
AFAIU having those utilities installed in `sanoid`'s package wrapper or `syncoid`'s package wrapper, would not help. Those utilities are called remotely by the local `syncoid`.

```console
$ nix path-info -rsSh $(which mbuffer) $(which lzop)
/nix/store/567zfi9026lp2q6v97vwn640rv6i3n4c-libunistring-1.1  	   1.8M	   1.8M
/nix/store/4563gldw8ibz76f1a3x69zq3a1vhdpz9-libidn2-2.3.4     	 350.4K	   2.1M
/nix/store/jd99cyc0251p0i5y69w8mqjcai8mcq7h-xgcc-12.2.0-libgcc	 139.3K	 139.3K
/nix/store/46m4xx889wlhsdj72j38fnlyyvvvvbyb-glibc-2.37-8      	  28.8M	  31.1M
/nix/store/4iabmjjq95069myjsrid8pk2ib3yz4nn-openssl-3.0.10    	   6.2M	  37.3M
/nix/store/sxrh37b22ga4q42kcapfh1746jk887ky-lzo-2.10          	 392.3K	  31.5M
/nix/store/6k4g5zdbddw4w72ahna4sz2fbb091xbr-lzop-1.04         	 234.8K	  31.7M
/nix/store/ffj3cdbza79b1lgd5z0ylg76fj0byk5w-mbuffer-20230301  	 114.2K	  37.4M
```

lopsided98 commented on nixos/modules/services/backup/syncoid.nix:
```diff
@@ -406,18 +350,28 @@ in
+            } //
+            (
+            if length sshKeyCred > 1
+            then { LoadCredentialEncrypted = [ c.sshKey ]; }
```
I can see how this is useful, particularly since using `LoadCredential` for the SSH key somewhat precludes configuring `LoadCredentialEncrypted` manually, but it seems odd that no other service has implemented a similar pattern. I guess the feature is too new?

ju1m replied:
You're right, I came up with it on my own. AFAICS it could become a pattern but is not yet.

lopsided98 reviewed (commented):
I haven't tested the latest version of this PR, but I think the changes make sense and look good overall. I just wish we could get someone with commit privileges to take an interest (the same goes for all the other sanoid/syncoid PRs).

NiklasGollenstede reviewed (commented):
I was directed here from/by https://github.com/NixOS/nixpkgs/pull/208121#issuecomment-1724570209 (thanks!).

Generally, I am a big fan of statelessness (and thus DynamicUser, etc).
I do though share the concern about the permission delegation to ephemeral users (https://github.com/NixOS/nixpkgs/pull/147559#pullrequestreview-887964206). (Unfortunately) ZFS needs this to be done in a stateful manner, which is inherently conflicting DynamicUser.
(Also unfortunately,) I don't think that the explanations in https://github.com/NixOS/nixpkgs/pull/147559#issuecomment-1046445837 are sufficient.
Systemd usually picking the same same UID for the same user is (as far as I can tell) _entirely_ implementation detail. The hashing stuff quite clearly is, and also the StateDirectory stuff is only to ensure that the service is able to access that directory. There are other ways to ensure that (dynamic filesystem UID remapping), and systemd might change to completely random UIDs in the future (without breaking any promised APIs). And (if not already) I could quite imagine that systemd will allow any system user to define and spawn DynamicUser units, giving them a decent chance to acquire a previously used ID.

Now, since ZFSs permissions (in this scenario) are only stateful in case of abrupt reboots, I could imagine a one-shot-at-boot service that removes all permissions issued to UIDs in systemd's dynamic UID range from all (potentially affected) datasets.
If that seems too brute-force, one could also save the last-used UID in a custom (service and dataset specific) ZFS property and only remove permissions issued to that UID (on boot or next main unit start).

lopsided98 commented on nixos/modules/services/backup/syncoid.nix:
```diff
@@ -138,21 +67,19 @@ in
     localTargetAllow = mkOption {
       type = types.listOf types.str;
-      default = [ "change-key" "compression" "create" "mount" "mountpoint" "receive" "rollback" ];
+      default = [ "change-key" "compression" "create" "destroy" "mount" "mountpoint" "receive" "rollback" ];
```
There is an existing discussion of this change in https://github.com/NixOS/nixpkgs/issues/216614, and I don't think anyone has brought up this point. I've always been wary of the destroy permission as well, but I had previously thought it was mostly harmless in `localTargetAllow` because I hadn't considered the pull backup case.

The main argument for making it part of the defaults is that syncoid's default configuration (without `--no-sync-snap`) has to destroy a snapshot each time it runs.

ju1m replied:
I was not aware of https://github.com/NixOS/nixpkgs/issues/216614. As described in 2035c7fcbc5b2d9ea1ffbca5246f498d9adc1ee0 (body of the commit message), adding `destroy` to `localTargetAllow` is required to reset broken receive states (`zfs receive -A`), which syncoid does when it fails to resume a receive state, because the snapshot has been destroyed on the source (which happened here to me on an `_hourly` sanoid snapshot which had been destroyed by a sanoid policy between the first run of syncoid which was interrupted for whatever reason, and the new one).

To confirm this, I just reproduced it like so:
```console
# Creating a source dataset, @file will be destroyed
root@losurdo$ zfs create das1/julm/tank
root@losurdo$ zfs snapshot das1/julm/tank@init
root@losurdo$ dd if=/dev/urandom of=/mnt/das1/julm/tank/file.rand status=progress count=10 bs=10MiB
root@losurdo$ zfs snapshot das1/julm/tank@file

# Creating a target dataset, `destroy` is not yet allowed to `julm`
root@oignon$ zfs create oignon/backup
root@oignon$ zfs allow julm change-key,compression,create,mount,mountpoint,receive,rollback oignon/backup
root@oignon$ chmod a+rwx /backup

# Testing the receiving
julm@oignon$ syncoid --no-privilege-elevation --no-sync-snap losurdo.wg:das1/julm/tank oignon/backup/tank
NEWEST SNAPSHOT: file
Sending incremental das1/julm/tank@init ... file (~ 100.2 MB):
^C

# Testing the resuming, and aborting again
julm@oignon$ syncoid --no-privilege-elevation --no-sync-snap losurdo.wg:das1/julm/tank oignon/backup/tank
Resuming interrupted zfs send/receive from das1/julm/tank to oignon/backup/tank (~ 100.1 MB remaining):
^C

# Destroying the source snapshot used by the receiving state
root@losurdo$ zfs destroy das1/julm/tank@file

# Trying to resume again: sending fails to `destroy` the receiving state
julm@oignon$ syncoid --debug --no-privilege-elevation --no-sync-snap losurdo.wg:das1/julm/tank oignon/backup/tank
DEBUG: SSHCMD: ssh     
DEBUG: checking availability of lzop on source...
DEBUG: checking availability of lzop on target...
DEBUG: checking availability of lzop on local machine...
DEBUG: checking availability of mbuffer on source...
DEBUG: checking availability of mbuffer on target...
DEBUG: checking availability of pv on local machine...
DEBUG: checking availability of zfs resume feature on source...
DEBUG: checking availability of zfs resume feature on target...
DEBUG: syncing source das1/julm/tank to target oignon/backup/tank.
DEBUG: getting current value of syncoid:sync on das1/julm/tank...
ssh      -S /tmp/syncoid-losurdo.wg-1696281307-7975 losurdo.wg  zfs get -H syncoid:sync ''"'"'das1/julm/tank'"'"''
DEBUG: checking to see if oignon/backup/tank on  is already in zfs receive using  ps -Ao args= ...
DEBUG: checking to see if target filesystem exists using "  zfs get -H name 'oignon/backup/tank' 2>&1 |"...
DEBUG: getting current value of receive_resume_token on oignon/backup/tank...
  zfs get -H receive_resume_token 'oignon/backup/tank'
DEBUG: got receive resume token: 1-de168d8de-e8-789c636064000310a501c49c50360710a715e5e7a69766a63040816a07d7572d29cdf50a40363b92bafca4acd4e412060616a83a0cf9b4b4e2d412063800c9b321c9275596a41603e90fbc2c58f597e4435c61794cd6e947d5ad670648f29c60f9bcc4dc54068694c46243fdacd29c5cfd92c4bc6c87b4cc9c54b89d00c9b12039: 
DEBUG: getting estimated transfer size from source -S /tmp/syncoid-losurdo.wg-1696281307-7975 losurdo.wg using "ssh      -S /tmp/syncoid-losurdo.wg-1696281307-7975 losurdo.wg  zfs send  -nvP -t 1-de168d8de-e8-789c636064000310a501c49c50360710a715e5e7a69766a63040816a07d7572d29cdf50a40363b92bafca4acd4e412060616a83a0cf9b4b4e2d412063800c9b321c9275596a41603e90fbc2c58f597e4435c61794cd6e947d5ad670648f29c60f9bcc4dc54068694c46243fdacd29c5cfd92c4bc6c87b4cc9c54b89d00c9b12039 2>&1 |"...
DEBUG: sendsize = 
Resuming interrupted zfs send/receive from das1/julm/tank to oignon/backup/tank (~ UNKNOWN remaining):
DEBUG: ssh      -S /tmp/syncoid-losurdo.wg-1696281307-7975 losurdo.wg ' zfs send  -t 1-de168d8de-e8-789c636064000310a501c49c50360710a715e5e7a69766a63040816a07d7572d29cdf50a40363b92bafca4acd4e412060616a83a0cf9b4b4e2d412063800c9b321c9275596a41603e90fbc2c58f597e4435c61794cd6e947d5ad670648f29c60f9bcc4dc54068694c46243fdacd29c5cfd92c4bc6c87b4cc9c54b89d00c9b12039 | lzop  | mbuffer  -q -s 128k -m 16M' | mbuffer  -q -s 128k -m 16M | lzop -dfc | pv -p -t -e -r -b -s 0 |  zfs receive  -s -F 'oignon/backup/tank' 2>&1
cannot resume send: 'das1/julm/tank@file' used in the initial send no longer exists
cannot receive: failed to read from stream
WARN: resetting partially receive state because the snapshot source no longer exists
DEBUG: reset partial receive state of oignon/backup/tank...
  zfs receive -A 'oignon/backup/tank'
cannot destroy 'oignon/backup/tank/%recv': permission denied
CRITICAL ERROR:   zfs receive -A 'oignon/backup/tank' failed: 256 at /nix/store/8qlzmagqra5vrkm80hm0lf16wg1095nx-sanoid-2.2.0/bin/.syncoid-wrapped line 2177.

# Allowing `julm` to `destroy`
root@oignon$ sudo zfs allow julm destroy oignon/backup/tank
# Now sending works again
julm@oignon$ syncoid --debug --no-privilege-elevation --no-sync-snap losurdo.wg:das1/julm/tank
DEBUG: SSHCMD: ssh     
DEBUG: checking availability of lzop on source...
DEBUG: checking availability of lzop on target...
DEBUG: checking availability of lzop on local machine...
DEBUG: checking availability of mbuffer on source...
DEBUG: checking availability of mbuffer on target...
DEBUG: checking availability of pv on local machine...
DEBUG: checking availability of zfs resume feature on source...
DEBUG: checking availability of zfs resume feature on target...
DEBUG: syncing source das1/julm/tank to target oignon/backup/tank.
DEBUG: getting current value of syncoid:sync on das1/julm/tank...
ssh      -S /tmp/syncoid-losurdo.wg-1696281928-2345 losurdo.wg  zfs get -H syncoid:sync ''"'"'das1/julm/tank'"'"''
DEBUG: checking to see if oignon/backup/tank on  is already in zfs receive using  ps -Ao args= ...
DEBUG: checking to see if target filesystem exists using "  zfs get -H name 'oignon/backup/tank' 2>&1 |"...
DEBUG: getting current value of receive_resume_token on oignon/backup/tank...
  zfs get -H receive_resume_token 'oignon/backup/tank'
DEBUG: got receive resume token: 1-de168d8de-e8-789c636064000310a501c49c50360710a715e5e7a69766a63040816a07d7572d29cdf50a40363b92bafca4acd4e412060616a83a0cf9b4b4e2d412063800c9b321c9275596a41603e90fbc2c58f597e4435c61794cd6e947d5ad670648f29c60f9bcc4dc54068694c46243fdacd29c5cfd92c4bc6c87b4cc9c54b89d00c9b12039: 
DEBUG: getting estimated transfer size from source -S /tmp/syncoid-losurdo.wg-1696281928-2345 losurdo.wg using "ssh      -S /tmp/syncoid-losurdo.wg-1696281928-2345 losurdo.wg  zfs send  -nvP -t 1-de168d8de-e8-789c636064000310a501c49c50360710a715e5e7a69766a63040816a07d7572d29cdf50a40363b92bafca4acd4e412060616a83a0cf9b4b4e2d412063800c9b321c9275596a41603e90fbc2c58f597e4435c61794cd6e947d5ad670648f29c60f9bcc4dc54068694c46243fdacd29c5cfd92c4bc6c87b4cc9c54b89d00c9b12039 2>&1 |"...

DEBUG: sendsize = 
Resuming interrupted zfs send/receive from das1/julm/tank to oignon/backup/tank (~ UNKNOWN remaining):
DEBUG: ssh      -S /tmp/syncoid-losurdo.wg-1696281928-2345 losurdo.wg ' zfs send  -t 1-de168d8de-e8-789c636064000310a501c49c50360710a715e5e7a69766a63040816a07d7572d29cdf50a40363b92bafca4acd4e412060616a83a0cf9b4b4e2d412063800c9b321c9275596a41603e90fbc2c58f597e4435c61794cd6e947d5ad670648f29c60f9bcc4dc54068694c46243fdacd29c5cfd92c4bc6c87b4cc9c54b89d00c9b12039 | lzop  | mbuffer  -q -s 128k -m 16M' | mbuffer  -q -s 128k -m 16M | lzop -dfc | pv -p -t -e -r -b -s 0 |  zfs receive  -s -F 'oignon/backup/tank' 2>&1
cannot resume send: 'das1/julm/tank@file' used in the initial send no longer exists
cannot receive: failed to read from stream
WARN: resetting partially receive state because the snapshot source no longer exists
DEBUG: reset partial receive state of oignon/backup/tank...
  zfs receive -A 'oignon/backup/tank'
DEBUG: syncing source das1/julm/tank to target oignon/backup/tank.
DEBUG: getting current value of syncoid:sync on das1/julm/tank...
ssh      -S /tmp/syncoid-losurdo.wg-1696281928-2345 losurdo.wg  zfs get -H syncoid:sync ''"'"'das1/julm/tank'"'"''
DEBUG: checking to see if oignon/backup/tank on  is already in zfs receive using  ps -Ao args= ...
DEBUG: checking to see if target filesystem exists using "  zfs get -H name 'oignon/backup/tank' 2>&1 |"...
DEBUG: getting current value of receive_resume_token on oignon/backup/tank...
  zfs get -H receive_resume_token 'oignon/backup/tank'
DEBUG: no receive token found 
DEBUG: getting list of snapshots on das1/julm/tank using ssh      -S /tmp/syncoid-losurdo.wg-1696281928-2345 losurdo.wg  zfs get -Hpd 1 -t snapshot guid,creation ''"'"'das1/julm/tank'"'"'' |...
DEBUG: getting list of snapshots on oignon/backup/tank using   zfs get -Hpd 1 -t snapshot guid,creation 'oignon/backup/tank' |...
NEWEST SNAPSHOT: init
DEBUG: getting current value of -p used on oignon/backup/tank...
  zfs get -H -p used 'oignon/backup/tank'
DEBUG: checking to see if oignon/backup/tank on  is already in zfs receive using  ps -Ao args= ...
INFO: no snapshots on source newer than init on target. Nothing to do, not syncing.
```

[zfs-allow(8)](https://openzfs.github.io/openzfs-docs/man/master/8/zfs-allow.8.html) doesn't seem to give a finer permission than `destroy`.

ju1m commented on nixos/modules/services/backup/syncoid.nix:
```diff
@@ -89,6 +26,16 @@ in
     package = lib.mkPackageOptionMD pkgs "sanoid" {};
 
+    nftables.enable = mkEnableOption (lib.mdDoc ''
+      nftables integration.
```
Thanks for the clarifying description.
I've not renamed `enable` to `maintainUidSet` though, because IIRC there's a pattern throughout NixOS where integration into other services is just an unspecific `enable`.

ju1m commented:
> I do though share the concern about the permission delegation to ephemeral users ([#147559 (review)](https://github.com/NixOS/nixpkgs/pull/147559#pullrequestreview-887964206)). (Unfortunately) ZFS needs this to be done in a stateful manner, which is inherently conflicting DynamicUser. (Also unfortunately,) I don't think that the explanations in [#147559 (comment)](https://github.com/NixOS/nixpkgs/pull/147559#issuecomment-1046445837) are sufficient.

They're not. I checked my test system which has had several crashed when syncoid or sanoid services were running, and somehow two unknown dynamic UIDs were still allowed.

> Now, since ZFSs permissions (in this scenario) are only stateful in case of abrupt reboots, I could imagine a one-shot-at-boot service that removes all permissions issued to UIDs in systemd's dynamic UID range from all (potentially affected) datasets. If that seems too brute-force, one could also save the last-used UID in a custom (service and dataset specific) ZFS property and only remove permissions issued to that UID (on boot or next main unit start).

I've added an `ExecStartPre=`s to each syncoid service calling `zfs-unallow-unused-dynamic-users` on local datasets. That bash script parses the output of `zfs allow` to remove the permissions given to any unused dynamic users in the allow list (ie. users without a name and an UID in the range 61184…65519).

The same should be done for the sanoid service. But I think that's already enough to get your opinions on this solution. Thanks a lot to the both of you for your interest in this PR and insightful comments.

NiklasGollenstede commented:
Sorry it took me some time to reply. I think some of my comments make more sense if I can refer to my own ZFS backup scheme, which I spent way too much time working out and refining over the last years, but which I think has reached relative maturity by now. So I took another significant chunk of time to refactor and document it before now making it public.

I want a backup scheme where compromised hosts can't poison any previously backups on the other hosts involved (e.g., an attacker-owned production server can't delete any backups made before, and an attacker-owned backup target can't affect the backup sources more than that host simply failing).

Most crucial for that is that SSH users (for which some other system has the SSH key) can not have the `destroy` permission (and I think this also applies to a large extend to a local target user when pulling, but that is not the scenario I use).
One side of the send/receive not being able to destroy any snapshots at all is quite a pain, especially if you want to support backup forwarding and hosts that do not reliably or permanently have internet.
Not needing to worry about any local users, at least, is nice, so I quite appreciate this PR.

I say this mostly relating to the side discussions we have here:
* remote host packages (https://github.com/NixOS/nixpkgs/pull/236729#discussion_r1329375132): I actually even force the remote user to run a [specific command](https://github.com/NiklasGollenstede/nix-wiplib/tree/master/modules/services/zfs/utils/ssh-command-filter.sh), which filters the remote syncoid's `SSH_ORIGINAL_COMMAND` and provides the necessary packages (`mbuffer`) to run it.
* `destroy` permission on the receiving side (https://github.com/NixOS/nixpkgs/pull/236729#discussion_r1329630660): I use `--no-sync-snap` or `--keep-sync-snap(-target)`, some patches to syncoid, and periodic cleanup jobs run locally on the target side to avoid giving the remote/receiving user the `destroy` permission.

If you are interested, and I would certainly appreciate comments and opinions, this is all explained in more detail and implemented [here](https://github.com/NiklasGollenstede/nix-wiplib/tree/master/modules/services/zfs/README.md).

NiklasGollenstede commented on nixos/modules/services/backup/syncoid.nix:
```diff
@@ -301,42 +227,103 @@ in
+                # For a local source, allow the localSourceAllow ZFS permissions.
+                map
+                  (dataset:
+                    "+/run/booted-system/sw/bin/zfs allow $USER " +
+                      escapeShellArgs [ (concatStringsSep "," c.localSourceAllow) dataset ]
+                  )
+                  (localDatasetName c.source) ++
+                # For a local target, check if the dataset exists before delegating permissions,
+                # and if it doesn't exist, delegate it to the parent dataset.
+                # This should solve the case of provisioning new datasets.
+                map
+                  (dataset:
+                    "+" + pkgs.writeShellScript "zfs-allow-target" ''
+                      dataset="$1"
+                      # Run a ZFS list on the dataset to check if it exists
+                      zfs list "$dataset" >/dev/null 2>/dev/null ||
+                        dataset="$(dirname "$dataset")"
+                      zfs allow "$USER" ${escapeShellArg (concatStringsSep "," c.localTargetAllow)} "$dataset"
+                    '' + " " + escapeShellArg dataset
+                  )
+                  (localDatasetName c.target) ++
+                # Adding a user to an nftables set will not persist across a reboot,
+                # hence there is no need to cleanup residual dynamic users remaining in it after a crash.
+                optional cfg.nftables.enable
+                  "+${pkgs.nftables}/bin/nft add element inet filter nixos-syncoid-uids { $USER }";
               ExecStopPost =
-                (map (buildUnallowCommand c.localSourceAllow) (localDatasetName c.source)) ++
-                (map (buildUnallowCommand c.localTargetAllow) (localDatasetName c.target));
+                let
+                  zfsUnallow = dataset: "+/run/booted-system/sw/bin/zfs unallow $USER " + escapeShellArg dataset;
+                in
+                map zfsUnallow (localDatasetName c.source) ++
+                  # For a local target, unallow both the dataset and its parent,
+                  # because at this stage we have no way of knowing if the allow command
+                  # did execute on the parent dataset or not in the ExecStartPre=.
+                  # We can't run the same if-then-else in the post hook
+                  # since the dataset should have been created at this point.
+                  concatMap
+                    (dataset: [ (zfsUnallow dataset) (zfsUnallow (builtins.dirOf dataset)) ])
+                    (localDatasetName c.target) ++
```
This logic (delegating to the parent for local targets) also seems good to me.
(It could use the same implementation for `ExecStopPost` as in line 282.)

ju1m commented on nixos/modules/services/backup/syncoid.nix:
```diff
@@ -301,42 +227,103 @@ in
+                        while read -r role id; do
+                          if [ "$id" -ge 61184 ] && [ "$id" -le 65519 ]; then
+                            case "$role" in
+                              (user) uids+=("$id");;
```
I was hesitating, I was tempted to also clean dynamic groups even though this `services.syncoid` does not use them so far.

ju1m commented on nixos/modules/services/backup/syncoid.nix:
```diff
@@ -301,42 +227,103 @@ in
+                      {
+                        declare -a uids
+                        while read -r role id; do
+                          if [ "$id" -ge 61184 ] && [ "$id" -le 65519 ]; then
```
Sounds like a good idea. I guess it would first deserve its own PR to expose `DYNAMIC_UID_MIN`/`MAX`, and then another PR to use them in `services.syncoid`.

lschuermann committed:
nixos/syncoid: allow interval to be list of strings

The syncoid module generates systemd services with a `startAt` config
based on its `interval` value. However, it is currently not possible
to completely disable this timer, or set multiple `startAt` values.

By accepting the same types that `systemd.services.<name>.startAt`
accepts, we can disable generating a timer unit or set multiple
`startAt` timers.

ju1m commented:
- [X] Rebase, fixing merge conflict due to treewide `nixfmt-rfc-style`
- [X] Use the new `NFTSet=` instead of calling `nft`

ju1m committed:
nixos/syncoid: format with nixfmt

ju1m committed:
nixos/sanoid: format with nixfmt

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/syncoid-questions/26411/3
